libia2/include/ia2.h

```diff
@@ -264,13 +254,86 @@ static int insecure_pkey_mprotect(void *ptr, size_t len, int prot, int pkey) {
+    uint32_t new_pkru = ~((3 << (2 * i)) | 3);                                 \
+    __asm__ volatile(IA2_WRPKRU : : "a"(new_pkru), "d"(0), "c"(0));            \
+    /* Forbid overwriting an existing stack. */                                \
+    if (ia2_stackptrs[stack_idx] != NULL) {                                    \
+      printf(                                                                  \
+          "compartment %zd in thread %zd tried to allocate existing stack\n",  \
+          i, tid);                                                             \
+      exit(1);                                                                 \
+    }                                                                          \
+    ia2_stackptrs[stack_idx] = stack_ptr_value;                                \
+    __asm__(IA2_WRPKRU : : "a"(old_pkru), "d"(0), "c"(0));                     \
+  }                                                                            \
+                                                                               \
+  size_t ia2_alloc_compacted_tid(void);                                        \
+                                                                               \
+  void init_stacks(void) {                                                     \
+    extern __thread size_t ia2_compacted_tid;                                  \
+    ia2_compacted_tid = ia2_alloc_compacted_tid();                             \
     for (int i = 0; i < n + 1; i++) {                                          \
-      char *stack_start = allocate_stack(i);                                   \
-      protect_stack(i, stack_start);                                           \
-      /* Each stack frame start + 8 is initially 16-byte aligned. */           \
-      ia2_stackptrs[i] = stack_start + STACK_SIZE - 8;                         \
+      allocate_compartment_stack(i);                                           \
     }                                                                          \
+  }                                                                            \
+                                                                               \
+  extern void ia2_init_threads(void);                                          \
+                                                                               \
+  __attribute__((constructor)) static void ia2_init(void) {                    \
+    /* set up global resources */                                              \
+    ia2_init_threads();                                                        \
+    ensure_pkeys_allocated(&ia2_n_pkeys_to_alloc);                             \
+    protect_stackptrs();                                                       \
+    /* initialize stacks for the main thread */                                \
+    init_stacks();                                                             \
```

Comment:
This call wouldn't need the wrpkru back to `old_pkru`, but maybe we can fix that in a follow-up

Reply:
This isn't exactly true: `init_stacks` allocates a stack for each compartment, in numerical order, so we can't just stop at the current compartment if it isn't the last one.

Reply:
Oh right we're calling wrpkru in a loop